service/sharddistributor/handler/executor.go

```diff
@@ -57,16 +57,74 @@ func (h *executor) Heartbeat(ctx context.Context, request *types.ExecutorHeartbe
 
 	heartbeatTime := h.timeSource.Now().UTC()
 	mode := h.migrationConfiguration.GetMigrationMode(request.Namespace)
+
+	// Handle draining executor separately
+	if request.Status == types.ExecutorStatusDRAINING {
+		return h.handleDrainingHeartbeat(ctx, request, heartbeatTime, mode)
+	}
+
+	return h.handleActiveHeartbeat(ctx, request, previousHeartbeat, assignedShards, heartbeatTime, mode)
+}
+
+// handleDrainingHeartbeat handles heartbeats from executors that are shutting down.
+// It clears their shard assignments, so they can be reassigned to other executors.
+func (h *executor) handleDrainingHeartbeat(
+	ctx context.Context,
+	request *types.ExecutorHeartbeatRequest,
+	heartbeatTime time.Time,
+	mode types.MigrationMode,
+) (*types.ExecutorHeartbeatResponse, error) {
+	h.logger.Info("Executor is draining, clearing shard assignments",
+		tag.ShardNamespace(request.Namespace),
+		tag.ShardExecutor(request.ExecutorID))
+
+	if err := h.storage.DeleteExecutors(ctx, request.Namespace, []string{request.ExecutorID}, store.NopGuard()); err != nil {
+		h.logger.Error("Failed to delete draining executor's assignments",
+			tag.ShardNamespace(request.Namespace),
+			tag.ShardExecutor(request.ExecutorID),
+			tag.Error(err))
+		// Continue anyway - the leader will eventually reassign the shards
+	}
+
+	drainingHeartbeat := store.HeartbeatState{
+		LastHeartbeat:  heartbeatTime,
+		Status:         request.Status,
+		ReportedShards: request.ShardStatusReports,
+		Metadata:       request.GetMetadata(),
+	}
+	if err := h.storage.RecordHeartbeat(ctx, request.Namespace, request.ExecutorID, drainingHeartbeat); err != nil {
```

Comment:
This method writes the reported status, so maybe it just works already now that it sends the drained status in the heartbeat. 

this line adds the compressed status to the transaction `clientv3.OpPut(stateKey, string(compressedState)),`